File: 1582-design-browser-history/design-browser-history.py
class BrowserHistory:

    def __init__(self, homepage: str):
        self.pages = []
        self.current = 0
        self.pages.append(homepage)
        
    def visit(self, url: str) -> None:
        while len(self.pages) - 1 != self.current:
            self.pages.pop(-1)
        self.pages.append(url)
        self.current+=1

    def back(self, steps: int) -> str:
        self.current = max(0,self.current - steps)
        return self.pages[self.current]


    def forward(self, steps: int) -> str:
        self.current = min(len(self.pages) - 1, self.current + steps)
        return self.pages[self.current]

    # SC -> O(N)
    # TC -> O(N)
        


# Your BrowserHistory object will be instantiated and called as such:
# obj = BrowserHistory(homepage)
# obj.visit(url)
# param_2 = obj.back(steps)
# param_3 = obj.forward(steps)

File: 1582-design-browser-history/design-browser-history.java
class BrowserHistory {
    List<String> pages;
    int current;

    public BrowserHistory(String homepage) {
        this.pages = new ArrayList<>();
        this.pages.add(homepage);
        this.current = 0;        
    }
    
    public void visit(String url) {
        while(this.pages.size() - 1 != this.current){
            this.pages.remove(this.pages.size() - 1);
        }

        this.pages.add(url);
        this.current+=1;
    }
    
    public String back(int steps) {
        this.current = Math.max(0,this.current - steps);
        return this.pages.get(this.current);
        
    }
    
    public String forward(int steps) {
        this.current = Math.min(this.pages.size() - 1,this.current + steps);
        return this.pages.get(this.current);
        
    }
}

// SC - O(N)
// TC -> O(N)

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory obj = new BrowserHistory(homepage);
 * obj.visit(url);
 * String param_2 = obj.back(steps);
 * String param_3 = obj.forward(steps);
 */